### PR TITLE
feat: add intents.redirect() method ✨

### PR DESCRIPTION
**Pull request by gregorylegarec**
~PRIORITY : VERY LOW~

This PR tries to solve the redirection point exposed in #201 and adds two methods to intents API, both used to easily redirect to a given document or a given doctype. The two added methods are:

* `getRedirectionURL(type, data)` which retrieves a redirection URL.
* `redirect(type, data)` which retrieves a redirection URL and performs the redirection.

### Example

```jsx
import Button from 'cozy-ui/react/Button'
// [...]
  render() {
    const { folder } = this.props
    return (
      <Button
        label={`Open ${folder.name}`}
        onClick={() => cozy.client.intents.redirect('io.cozy.files', folder)}
      />
    )
  }
```

### Declaring a redirection handler

Redirection are declared in `manfiest.webapp`, like a regular intent

```json
"intents": [
    {
      "action": "REDIRECT",
      "type": [
        "io.cozy.konnectors"
      ],
      "href": "/"
    }
  ],
```

### Current limitations
* For now, it is still not possible to remove an intent from database.
* Statck cannot handle an intent having a sharp character in this `href` attribute, and URLencode it (i.e. replaces `#` by `%23`).

### CODE REVIEW

File: src/intents.js [hunk 1]
```diff
@@ -108,6 +108,16 @@ function injectService (url, element, intent, data, onReadyCallback) {
 }
 
 const first = arr => arr && arr[0]
+// In a far future, the user will have to pick the desired service from a list.
+// For now it's our job, an easy job as we aribtrary pick the first service of
```
Comment (CPatchane) [inline src/intents.js]: typo s/aribtrary/arbitrary

File: src/intents.js [hunk 2]
```diff
@@ -108,6 +108,16 @@ function injectService (url, element, intent, data, onReadyCallback) {
...
+// the list.
+function pickService (intent, filterServices) {
```
Comment (CPatchane) [inline src/intents.js]: 👍

**Comment by CPatchane** [comment]
For the `#` in the URL, maybe we can add an exception? Same thing for the `?` and `&` characters since they can be used in a URL? Just a suggestion, maybe it's not necessary for now.

File: src/intents.js (hunk 1 above)
Comment (y-lohse) [reply src/intents.js]: Isn't that comment out of date though ? The first service is the default choice now, but the `filterServices` already allows you to pick one, right?

File: src/intents.js [hunk 3]
```diff
@@ -234,3 +238,40 @@ export function createService (cozy, intentId, serviceWindow) {
...
+  const baseURL = service.href.split('?')[0]
+  return doc ? parseRedirectionURL(baseURL, doc) : baseURL
+}
+
+function parseRedirectionURL (url, doc) {
+  // replace /:params/ segments
+  return url.replace(/:([^//]+)/g, segment => {
+    const param = segment.substring(1)
+    if (!doc[param]) throw new Error(`Cannot generate redirection URL, document misses ${param} attribute.`)
+    return doc[param]
```
Comment (y-lohse) [inline src/intents.js]: Neat 👍

File: src/intents.js (hunk 2 above)
Comment (ptbrowne) [reply src/intents.js]: s/aribtrary/arbitrarily/

File: src/intents.js (hunk 1 above)
Comment (gregorylegarec) [reply src/intents.js]: As the `filterServices` is a callback given as parameter, it's not the user which pick the service, the choice is still made by code.

**Comment by gregorylegarec** [comment]
@CPatchane Actually, it's the stack wich generate the intent `href` and replace `#` by `%23` into it. I am not sure where it can be changed.

**Comment by CPatchane** [comment]
I think we could merge the two actions to use only a `REDIRECT` action and then pass parameters to be handled in the targeted application.

**Comment by gregorylegarec** [comment]
After meeting with @CPatchane, we agreed on improving the current implementation by using only a `REDIRECT` action and adding parameters in query string. This implies a better and explicit redirection handling in target application, but is more flexible and allow more possibilities.

The `manifest.webapp` declaration could evolve to enumerate the allowed parameters, first as information to contributor and developer, and maybe in the future for an automated use by the stack.

Example:

```json
"intents": [
    {
      "action": "REDIRECT",
      "type": [
        "io.cozy.konnectors"
      ],
      "href": "/#/discover",
      "params": ["slug, "category", "dataType"]
    }
  ],
```

**Comment by gregorylegarec** [comment]
[(3) ETU je peux filtrer dans mon Store](https://trello.com/c/HK3JrzHB/917-3-etu-je-peux-filtrer-dans-mon-store)

**Comment by CPatchane** [comment]
A fix has been added to handle the `#` character but must be removed when it will be correctly handled by the stack (ping @nono)

**Pull request merged**